search-provider/yelp-search-provider.c

```diff
@@ -0,0 +1,589 @@
+        g_variant_builder_add_value (&metas, g_variant_builder_end (&meta));
+    }
+
+    return g_variant_new ("(aa{sv})", &metas);
+}
+
+/* FIXME: Search was copied from yelp-search-entry and slightly adapted.
+ * When search in yelp gets adjusted we should export it and use it here.
+ */
+static gboolean
+match_word (const gchar *search_phrase,
+            const gchar *title_casefold,
+            const gchar *desc_casefold)
+{
+    static GRegex *nonword = NULL;
+    gint stri;
+    gchar **strs;
+    gboolean ret = TRUE;
+
+    if (nonword == NULL)
+        nonword = g_regex_new ("\\W", 0, 0, NULL);
+    g_assert (nonword != NULL);
+
+    strs = g_regex_split (nonword, search_phrase, 0);
+    for (stri = 0; strs[stri]; stri++) {
+        if (!title_casefold ||
+            !strstr (title_casefold, strs[stri]) &&
+            !desc_casefold ||
+            !strstr (desc_casefold, strs[stri])) {
+                ret = FALSE;
+                break;
+            }
+        }
+
+    g_strfreev (strs);
+
+    return ret;
+}
+
+static GVariant *
+get_search_results (gchar      **terms,
+                    GHashTable  *page_data)
+{
+    GVariantBuilder results;
+    gchar *term = g_strjoinv (" ", terms);
+    GHashTableIter iter;
+    const gchar *page_id;
+    PageData *data;
+
+    g_hash_table_iter_init (&iter, page_data);
+    g_variant_builder_init (&results, G_VARIANT_TYPE ("as"));
+
+    while (g_hash_table_iter_next (&iter, (gpointer *) &page_id, (gpointer *) &data)) {
+        if (match_word (term, data->title_casefold, data->desc_casefold))
+            g_variant_builder_add (&results, "s", page_id);
+    }
+
+    g_free (term);
+
+    return g_variant_new ("(as)", &results);
+}
+
+typedef GVariant * (*ResultGetter) (gchar      **terms,
+                                    GHashTable  *page_data);
+
+typedef struct
+{
+    GDBusMethodInvocation  *invocation;
+    gchar                 **terms;
+    ResultGetter            result_getter;
+} DelayedResultGetter;
+
+
+static DelayedResultGetter *
+delayed_result_getter_new (GDBusMethodInvocation  *invocation,
+                           gchar                 **terms,
+                           ResultGetter            result_getter)
+{
+    DelayedResultGetter *delayed = g_new0 (DelayedResultGetter, 1);
+
+    delayed->invocation = g_object_ref (invocation);
+    delayed->terms = g_strdupv (terms);
+    delayed->result_getter = result_getter;
+
+    return delayed;
+}
+
+static void
+delayed_result_getter_free (DelayedResultGetter *delayed)
+{
+    g_object_unref (delayed->invocation);
+    g_strfreev (delayed->terms);
+    g_free (delayed);
+}
+
+static gboolean 
+handle_get_initial_result_set (YelpShellSearchProvider2  *skeleton,
+                               GDBusMethodInvocation     *invocation,
+                               gchar                    **terms,
+                               gpointer                   user_data)
```

Comment:
It looks like the body of this function and `handle_get_subsearch_result_set()` are the same... Can you factor out a common method?